security/pkg/server/ca/server.go

```diff
@@ -81,38 +81,22 @@ func (s *Server) CreateCertificate(ctx context.Context, request *pb.IstioCertifi
 	crMetadata := request.Metadata.GetFields()
 	certSigner := crMetadata[security.CertSigner].GetStringValue()
 	log.Debugf("cert signer from workload %s", certSigner)
-	_, _, certChainBytes, rootCertBytes := s.ca.GetCAKeyCertBundle().GetAll()
+	_, _, _, rootCertBytes := s.ca.GetCAKeyCertBundle().GetAll()
 	certOpts := ca.CertOpts{
 		SubjectIDs: caller.Identities,
 		TTL:        time.Duration(request.ValidityDuration) * time.Second,
 		ForCA:      false,
 		CertSigner: certSigner,
 	}
-	cert, signErr := s.ca.Sign([]byte(request.Csr), certOpts)
+	respCertChain, signErr := s.ca.SignWithCertChain([]byte(request.Csr), certOpts)
 	if signErr != nil {
 		serverCaLog.Errorf("CSR signing error (%v)", signErr.Error())
 		s.monitoring.GetCertSignError(signErr.(*caerror.Error).ErrorType()).Increment()
 		return nil, status.Errorf(signErr.(*caerror.Error).HTTPErrorCode(), "CSR signing error (%v)", signErr.(*caerror.Error))
 	}
-	respCertChain := []string{string(cert)}
-	if len(certChainBytes) != 0 {
-		respCertChain = append(respCertChain, string(certChainBytes))
-	}
-	if len(rootCertBytes) == 0 {
-		rootCert, err := util.FindRootCertFromCertificateChainBytes(cert)
-		if err != nil {
-			serverCaLog.Errorf("failed to find root cert from signed cert-chain (%v)", err.Error())
-			s.monitoring.GetCertSignError(err.(*caerror.Error).ErrorType()).Increment()
-			return nil, status.Errorf(err.(*caerror.Error).HTTPErrorCode(), "failed to find root cert from signed cert-chain (%v)", err.(*caerror.Error))
-		}
-		if verifyErr := util.VerifyCertificate(nil, cert, rootCert, nil); verifyErr != nil {
-			serverCaLog.Errorf("root cert from signed cert-chain is invalid (%v)", err.Error())
-			s.monitoring.GetCertSignError(err.(*caerror.Error).ErrorType()).Increment()
-			return nil, status.Errorf(err.(*caerror.Error).HTTPErrorCode(), "root cert from signed cert-chain is invalid (%v)", err.(*caerror.Error))
-		}
-		rootCertBytes = rootCert
+	if len(rootCertBytes) != 0 {
```

Comment:
nit: is this still needed if the function invoked is "SignWithCertChain" as opposed to "Sign". Can this be handled inside the function

Reply:
yes, fixed